### PR TITLE
update convective operators for the advection-diffusion solver

### PR DESCRIPTION
**Pull request by boyceg**
### IBAMR Pull Request Checklist
- [x] Does the test suite pass?
- [x] Was clang-format run?
- [x] Were relevant issues cited? Please remember to add `Fixes #12345` to close
      the issue automatically if we are fixing the problem.
- [x] Is this a change others will want to know about? If so, then has a
      changelog entry been added?
- [x] If new data structures have been added to a class then have they been set
      up appropriately for restarts? If so, ensure that the restart version
      number is incremented.
- [x] Does this change include a bug fix or new functionality? If so, a new test
      or tests should be added. New tests should run quickly (less than a minute
      in release mode). If possible, an older test should gain a new option so
      that we do not need to compile more test executables.
- [x] Did you (if your account has permission to do so) set relevant labels on
      GitHub for the pull request?

One extra check: this needs to restore support for `SKEW_SYMMETRIC` differencing:
- [x] Restore support for `SKEW_SYMMETRIC` differencing!
- [x] Add checks that the flux patch data are allocated (or always allocate it).

### CODE REVIEW

**Comment by boyceg** [comment]
This is some code that I started on my flight on Saturday. I haven't tested it yet, and so there are likely bugs. But this is how I was thinking about making the convective operators more flexible, so that the CUI operator subclass in #1569 is not needed.

**Comment by boyceg** [comment]
OK, this now uses the new base class for centered, PPM, and CUI operators. Tests pass for me locally.

**Review by drwells — APPROVED** [review]
This looks good. Someone who uses advection-diffusion should review this further.

**Comment by boyceg** [comment]
I'd like to get some feedback from @abarret, @amneetb, and @Raamkrishnan on this before moving forward. For instance, I wonder if there are better choices for the method names. I also am not sure if the BC routines should be where they are here, or if they should go into the concrete implementations of the base class.

@amneetb, @Raamkrishnan: note that this functionality is intended to allow you to avoid needing to create the specialization of `AdvDiffCUIConvectiveOperator` that was introduced in the phase change branch.

**Comment by amneetb** [comment]
@Raamkrishnan please take a look and see if this suffices for the new operator. We can discuss offline when you have reviewed.

**Comment by amneetb** [comment]
@boyceg: @Raamkrishnan and I looked into the new `CellConvectiveOperator` class, but we can't see how it would do the $\nabla \cdot (\textbf{u} P Q)$ case? We see that it is designed to do either $\nabla \cdot (\textbf{u} P)$ or $\nabla \cdot (\textbf{u} Q)$. The only way this could be used in the phase change branch is that if we multiply $P$ and $Q$ together and limit the product. What we want to do is limit $P$ and $Q$ separately and then multiply them in the advective flux. Because $\text{limit}(P*Q) \ne \text{limit}(P)*\text{limit}(Q)$, we can't use this PR for the phase change branch unless we multiply P and Q together.

**Comment by amneetb** [comment]
@boyceg : We can try to multiply $P$ and $Q$ and then limit($PQ$) for a test problem and see if it makes a noticeable difference. I'm not sure what is done in practice in the CFD literature.

 @knepley : Do you have any recommendation for  $\nabla \cdot (\textbf{u} P Q)$? Should $P$ and $Q$ be limited separately and then multiplied or the product $PQ$ should be limited?

**Comment by boyceg** [comment]
The code here lets you separately interpolate each quantity to the faces (using whatever limiters are used in the convective operator implementation), then multiply them and take the divergence.

**Comment by amneetb** [comment]
But it means that we should not be using the default applyOperator() implementation, as it does things in a pre-set way.

**Comment by boyceg** [comment]
Correct, do not use applyOperator — use the interpolation and divergence functions instead.

**Comment by knepley** [comment]
@amneetb I have not read anything that argues that they should be limited separately, but I will check

**Comment by boyceg** [comment]
Note in particular that this code allows you either approach (limit and then multiply the face-centered fields to get the flux, or multiply the fields and then limit from cell centers to cell faces) without requiring a lot of code. E.g. the regular conservative differencing is just two function calls:
```
evaluateAdvectiveFluxOnHierarchy(d_q_flux_idx, Q_idx, d_u_idx, /*synch_cf_bdry*/ false);
computeConservativeDerivativeOnHierarchy(N_idx, d_q_flux_idx, /*synch_cf_bdry*/ true);
```
I don't know this literature, however, and if there is only way that we should ever be computing $\nabla \cdot {\bf f} = \nabla \cdot \left( \prod_k Q_k {\bf u} \right)$, then I am certainly happy to make that part of the interface.

File: include/ibamr/CellConvectiveOperator.h
Comment (abarret) [inline include/ibamr/CellConvectiveOperator.h:89]: Should these be marked `const`?
```suggestion
                                          const SAMRAI::pdat::CellData<NDIM, double>& Q_cell_data,
                                          const SAMRAI::pdat::FaceData<NDIM, double>& u_data,
```

File: include/ibamr/CellConvectiveOperator.h
Comment (abarret) [inline include/ibamr/CellConvectiveOperator.h:100]: Do we need to return data interpolated to the cell faces? At least in this implementation, we only need the flux.

File: include/ibamr/CellConvectiveOperator.h
Comment (abarret) [inline include/ibamr/CellConvectiveOperator.h:102]: Should these be marked `const`?
```suggestion
                                              const SAMRAI::pdat::CellData<NDIM, double>& Q_cell_data,
                                              const SAMRAI::pdat::FaceData<NDIM, double>& u_data,
```

**Review by abarret — COMMENTED** [review]
A few comments. I don't have any major comments on the interface, and the function names look good. If it works for @amneetb, then it works for me too.

Once the interface is finalized, we should update the documentation on the expected inputs and outputs of the added interface. I'd be happy to do this once everything else is finalized.

File: include/ibamr/CellConvectiveOperator.h
Comment (boyceg) [reply include/ibamr/CellConvectiveOperator.h:100]: The main rationale for requiring both is to avoid allocating/deallocating the temporary data for storing the `q_interp`. I can drop it from this interface, so that this function mirrors the interface of `evaluateAdvectiveFluxOnHierarchy`.

**Comment by boyceg** [comment]
@drwells: GitHub Actions is complaining:
```
Node.js 12 actions are deprecated. Please update the following actions to use Node.js 16: actions/checkout@v2. For more information see: https://github.blog/changelog/2022-09-22-github-actions-all-actions-will-begin-running-on-node16-instead-of-node12/.
```

**Comment by drwells** [comment]
Fortunately it is only deprecated - everything still works and the CI is failing for other reasons.

We need to get in these big patches before we do a bigger CI update but I think I can fix this one quickly.

File: include/ibamr/CellConvectiveOperator.h
Comment (boyceg) [reply include/ibamr/CellConvectiveOperator.h]: Fixed.

**Comment by drwells** [comment]
Want to add a changelog entry?

**Comment by drwells** [comment]
What needs to happen next here?

**Comment by boyceg** [comment]
Waiting on feedback from @Raamkrishnan and @amneetb.

**Comment by amneetb** [comment]
@Raamkrishnan Can you rebase the phase change branch against this one and do the requisite changes? Let's get this done first, before the regridding PR.

**Comment by Raamkrishnan** [comment]
> @Raamkrishnan Can you rebase the phase change branch against this one and do the requisite changes? Let's get this done first, before the regridding PR.

Sure. I will start working on this first.

**Comment by boyceg** [comment]
@drwells, @abarret, @amneetb: let me know if there are other changes needed.

**Comment by amneetb** [comment]
@Raamkrishnan will start using this PR to convert the convective operators introduced in the phase change branch next week. We will let you know if we run into some issues with this implementation.

**Comment by amneetb** [comment]
@Raamkrishnan will start using this PR to convert the convective operators introduced in the phase change branch next week. We will let you know if we run into some issues with this implementation.

**Review by drwells — APPROVED** [review]

**Comment by drwells** [comment]
@Raamkrishnan Did you get a chance to try this?

**Comment by Raamkrishnan** [comment]
@drwells I have not tried this yet. This week I will work on this.

**Comment by drwells** [comment]
Did you get a chance to check this out?

**Comment by Raamkrishnan** [comment]
> Did you get a chance to check this out?

I was busy with my rebuttal submission to one of our papers. I am done with that. I will be able to finish this week.

**Comment by drwells** [comment]
I'd like to get this in 0.13 but it isn't necessary (we only need this for upcoming patches which didn't make it into the release). Lets remove it from the milestone since it might not be ready.

**Comment by Raamkrishnan** [comment]
@drwells @boyceg @amneetb I have tested this in one of my phase change examples. The code can be found in 

```(https://github.com/IBAMR/IBAMR/blob/a3a2b43358d75b0b2f061de1350deb94166a296f/src/phase_change/AllenCahnHierarchyIntegrator.cpp#L458)```

It is working fine. Please take a look. If you do not have any more suggestions, we can merge this.

**Review by amneetb — APPROVED** [review]

**Pull request merged**
